### PR TITLE
Fixed bug in `dolt checkout` with uncommitted changes on another branch

### LINKED ISSUE #5725: Changes on multiple branches from sql-server has bad interaction with the CLI
Repro.

Step 1: Initialize a dolt database and create a table.
```
$ pwd
/Users/timsehn/dolthub/test_branch_changes
$ dolt init
Successfully initialized dolt data repository.
$ dolt sql -q "create table t (pk int primary key, c1 int, c2 int)"
$ dolt add t                                                       
$ dolt commit -am "Added table t"                                  
commit 1a6889bkrvueovlsb5vsdlao1l317gm6 (HEAD -> main) 
Author: timsehn <tim@dolthub.com>
Date:  Tue Apr 11 09:36:04 -0700 2023

        Added table t
```

Step 2: Start a SQL Server
```
$ dolt sql-server                                                  
Starting server with Config HP="localhost:3306"|T="28800000"|R="false"|L="info"|S="/tmp/mysql.sock"
2023-04-11T09:36:12-07:00 INFO [conn 1] NewConnection {DisableClientMultiStatements=false}
2023-04-11T09:37:23-07:00 INFO [conn 1] ConnectionClosed {}
```

Step 3: Connect to the SQL Server and Make changes on two branches
```
$ mysql -u root
Welcome to the MySQL monitor.  Commands end with ; or \g.
Your MySQL connection id is 1
Server version: 5.7.9-Vitess 

Copyright (c) 2000, 2023, Oracle and/or its affiliates.

Oracle is a registered trademark of Oracle Corporation and/or its
affiliates. Other names may be trademarks of their respective
owners.

Type 'help;' or '\h' for help. Type '\c' to clear the current input statement.

mysql> use test_branch_changes
Reading table information for completion of table and column names
You can turn off this feature to get a quicker startup with -A

Database changed
mysql> insert into t values (1,1,1);
Query OK, 1 row affected (0.02 sec)

mysql> call dolt_checkout('-b', 'branch');
+--------+
| status |
+--------+
|      0 |
+--------+
1 row in set (0.03 sec)

mysql> insert into t values (0,0,0);
Query OK, 1 row affected (0.02 sec)

mysql> exit
Bye
```

Step 4: Stop the SQL Server and inspect the database on the CLI:
```
$ dolt branch
  branch                                            
* main                                              
$ dolt status
On branch main
Changes not staged for commit:
  (use "dolt add <table>" to update what will be committed)
  (use "dolt checkout <table>" to discard changes in working directory)
    modified:       t
$ dolt checkout branch
error: There are uncommitted changes already on branch 'branch'.This can happen when someone modifies that branch in a SQL session.You have uncommitted changes on this branch, and they would overwrite the uncommitted changes on branch branch on checkout.To solve this problem, you can 1) commit or reset your changes on this branch, using `dolt commit` or `dolt reset`, before checking out the other branch, 2) use the `-f` flag with `dolt checkout` to force an overwrite, or 3) connect to branch '%s' with the SQL server and revert or commit changes there before proceeding.
cause: checkout would overwrite uncommitted changes on target branch
$ dolt branch
  branch                                            
* main    
$
```

Step 5: Commit changes to switch branches
```
$ dolt commit -am "added values"
commit t6fscukb0ptifen3e14kdb6koqi60dbe (HEAD -> main) 
Author: timsehn <tim@dolthub.com>
Date:  Tue Apr 11 09:38:18 -0700 2023

        added values

```

Step 6: Change branches and now branch changes are gone.
```
$ dolt checkout branch
Switched to branch 'branch'
$ dolt status
On branch branch
nothing to commit, working tree clean
$ dolt sql -q "select * from t"    

$ 
```

### PR DESCRIPTION
**Pull request by zachmu**
Fixes #5725

### CODE REVIEW

File: go/libraries/doltcore/env/actions/branch.go [hunk 1]
```diff
@@ -575,53 +605,55 @@ func overwriteRoot(ctx context.Context, head *doltdb.RootValue, tblHashes map[st
...
 
-// checkWorkingSetCompatibility checks that the current working set is "compatible" with the dest working set.
+// checkoutWouldStompWorkingSetChanges checks that the current working set is "compatible" with the dest working set.
 // This means that if both working sets are present (ie there are changes on both source and dest branches),
 // we check if the changes are identical before allowing a clobbering checkout.
 // Working set errors are ignored by this function, because they are properly handled elsewhere.
-func checkWorkingSetCompatibility(ctx context.Context, dEnv *env.DoltEnv, branchRef ref.BranchRef, currentWs *doltdb.WorkingSet) error {
-	db := dEnv.DoltDB
-	destWsRef, err := ref.WorkingSetRefForHead(branchRef)
+func checkoutWouldStompWorkingSetChanges(ctx context.Context, dEnv *env.DoltEnv, branchRef ref.BranchRef) bool {
```
Comment (PavelSafronov) [inline go/libraries/doltcore/env/actions/branch.go:612]: Is swallowing errors here problematic? Or are we relying on subsequent code to break?
Might be safer to return `true` (as that will cause the calling code to report an error), but then we'll end up reporting the wrong error to the user.

**Review by PavelSafronov — APPROVED** [review]

**Comment by timsehn** [comment]
These two bats tests fail locally:
```
 ✗ 1pk5col-ints: make a change on a different branch, commit, and merge to main
   (in test file integration-tests/bats/1pk5col-ints.bats, line 359)
     `[ "$status" -eq 0 ]' failed
   Successfully initialized dolt data repository.
   Switched to branch 'test-branch'
   Query OK, 1 row affected (0.00 sec)
   commit hkh96mid4tq4mqrudp54i6j17e9csvid (HEAD -> test-branch) 
   Author: Bats Tests <bats@email.fake>
   Date:  Thu Apr 13 10:22:12 -0700 2023
   
   	added test row
   
   Switched to branch 'main'

```
 ✗ 1pk5col-ints: display correct merge stats
   (in test file integration-tests/bats/1pk5col-ints.bats, line 736)
     `dolt merge test-branch --no-commit' failed
   Successfully initialized dolt data repository.
   Switched to branch 'test-branch'
   commit 3jdvhkl0fmk6qfdkt1aanorcfdvva3f4 (HEAD -> test-branch) 
   Author: Bats Tests <bats@email.fake>
   Date:  Thu Apr 13 10:22:40 -0700 2023
   
   	added test table
   
   Switched to branch 'main'
   Switched to branch 'test-branch-m'
   Updating bvqu56n3lf5h5679ng5753j58l0b4e6c..3jdvhkl0fmk6qfdkt1aanorcfdvva3f4
   error: Failed to commit changes.
   cause: error: Your local changes to the following tables would be overwritten by merge:
          test
          Please commit your changes before you merge.
```

Can't seem to repro the first failure outside of bats.

**Comment by timsehn** [comment]
So here is the issue, and I think it's an actual bug in "working sets following the checkout". Both of these tests kind of test this behavior indirectly. A table is created on main, making a working set. There is no commit and we create a new branch. The HEAD of the branch is the init commit. The created table is in the working set and follows the checkout to the new branch. We then checkout main and try and merge and now the merge fails with:
```
   error: Failed to commit changes.
   cause: error: Your local changes to the following tables would be overwritten by merge:
          test
          Please commit your changes before you merge.
```

**Comment by timsehn** [comment]
This simple repro shows two bugs with working sets following checkout on the command line with this change:
```
$ dolt init 
Successfully initialized dolt data repository.
$ dolt sql -q "create table t (pk int primary key, c1 int, c2 int)"
$ dolt checkout -b branch
Switched to branch 'branch'
$ dolt sql -q "insert into t values (0,0,0)"                       
Query OK, 1 row affected (0.00 sec)
$ dolt checkout main 
error: There are uncommitted changes already on branch 'main'.This can happen when someone modifies that branch in a SQL session.You have uncommitted changes on this branch, and they would overwrite the uncommitted changes on branch main on checkout.To solve this problem, you can 1) commit or reset your changes on this branch, using `dolt commit` or `dolt reset`, before checking out the other branch, 2) use the `-f` flag with `dolt checkout` to force an overwrite, or 3) connect to branch '%s' with the SQL server and revert or commit changes there before proceeding.
cause: checkout would overwrite uncommitted changes on target branch
$ dolt add t                                                       
$ dolt commit -m "committing changes"
commit 99r60kj84qatmnkc3ubo40tp2d5gna1a (HEAD -> branch) 
Author: timsehn <tim@dolthub.com>
Date:  Thu Apr 13 10:54:49 -0700 2023

        committing changes

$ dolt checkout main
Switched to branch 'main'
$ dolt merge branch                  
Updating 5gqn01f4b46a9tt6bvsp4cj4vohj2630..99r60kj84qatmnkc3ubo40tp2d5gna1a
error: Failed to commit changes.
cause: error: Your local changes to the following tables would be overwritten by merge:
       t
       Please commit your changes before you merge.
$
```

**Review by timsehn — CHANGES_REQUESTED** [review]

File: go/libraries/doltcore/env/actions/branch.go [hunk 2]
```diff
@@ -438,20 +477,11 @@ func CheckoutBranch(ctx context.Context, dEnv *env.DoltEnv, brName string, force
 		return err
 	}
 
-	if shouldResetWorkingSet {
```
[the remark this answers is not in the document]
Comment (zachmu) [reply go/libraries/doltcore/env/actions/branch.go:441]: Yes, I think this is the bug

[the remark this answers is not in the document]
Comment (zachmu) [reply go/libraries/doltcore/env/actions/branch.go:441]: It was actually more subtle than that: DoltReset doesn't touch new tables. Have to also DoltClean. This bug was masked by the buggy behavior I originally fixed.

**Review by timsehn — APPROVED** [review]

File: go/libraries/doltcore/env/actions/branch.go (hunk 1 above)
Comment (zachmu) [reply go/libraries/doltcore/env/actions/branch.go:679]: It kind of is, but if there's actually a persistent problem loading any of these entities we wouldn't even get this far.

**Pull request merged**